File: rust/src/problems/p51.rs
pub struct Solution;

impl Solution {
    pub fn solve_n_queens(n: i32) -> Vec<Vec<String>> {
        fn is_valid(queens: &[usize], col: usize) -> bool {
            queens.iter().all(|&c| c != col)
                && queens.iter().rev().enumerate().all(|(i, &c)| {
                    let c = c as i32;
                    let col = col as i32;
                    let i = i as i32;
                    c != col - i - 1 && c != col + i + 1
                })
        }

        fn dfs(queens: &mut Vec<usize>, row: usize, n: usize, result: &mut Vec<Vec<String>>) {
            if row == n {
                let mut board = Vec::with_capacity(n);
                for i in queens {
                    let mut row = vec!['.'; n];
                    row[*i] = 'Q';
                    board.push(row.iter().collect::<String>());
                }
                result.push(board);
                return;
            }
            for col in 0..n {
                if is_valid(queens, col) {
                    queens.push(col);
                    dfs(queens, row + 1, n, result);
                    queens.pop();
                }
            }
        }

        let n = n as usize;
        let mut result = vec![];
        dfs(&mut Vec::with_capacity(n), 0, n, &mut result);
        result
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn case1() {
        let expected = [
            [".Q..", "...Q", "Q...", "..Q."],
            ["..Q.", "Q...", "...Q", ".Q.."],
        ];
        let mut expected = expected
            .iter()
            .map(|b| b.iter().map(|r| r.to_string()).collect::<Vec<_>>())
            .collect::<Vec<_>>();
        expected.sort_unstable();
        let mut result = Solution::solve_n_queens(4);
        result.sort_unstable();
        assert_eq!(expected, result);
    }
}

File: cpp/p51.cpp
#include <algorithm>
#include <cassert>
#include <string>
#include <vector>

using namespace std;

class Solution {
    int n;
    vector<vector<string>> r;

public:
    bool is_valid(vector<int> &queens, int col) {
        int row = queens.size(), i2 = row + col, j2 = row - col;
        for (int i = 0; i < queens.size(); i++) {
            int j = queens[i];
            int i1 = i + j, j1 = i - j;
            if (j == col || i1 == i2 || j1 == j2) {
                return false;
            }
        }
        return true;
    }

    void helper(const int i, vector<int> &queens) {
        if (i == n) {
            vector<string> board(n, string(n, '.'));
            for (int i = 0; i < n; i++) {
                board[i][queens[i]] = 'Q';
            }
            r.push_back(board);
            return;
        }
        for (int j = 0; j < n; j++) {
            if (is_valid(queens, j)) {
                queens.push_back(j);
                helper(i + 1, queens);
                queens.pop_back();
            }
        }
    }

    vector<vector<string>> solveNQueens(int n) {
        this->n = n;
        vector<int> queens;
        queens.reserve(n);
        helper(0, queens);
        return r;
    }
};

int main() {
    vector<vector<string>> expected = {{".Q..", "...Q", "Q...", "..Q."},
                                       {"..Q.", "Q...", "...Q", ".Q.."}};
    sort(expected.begin(), expected.end());
    auto r = Solution().solveNQueens(4);
    sort(r.begin(), r.end());
    assert(r == expected);
    return 0;
}
